### PR TITLE
Adding Base structure for pod/container config

### PR DESCRIPTION
**Pull request by CharlyF**
Goal is to have

```yaml
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  credentials:
    apiKey: <DATADOG_API_KEY>
    appKey: <DATADOG_APP_KEY>
  features:
    orchestratorExplorer:
      enabled: true
    kubeStateMetricsCore:
      enabled: true
    apm:
      enabled: true
    dogstastd:
      enabled: true
  global: # global settings 
    clusterName: foo
    registry: gcr.io/datadoghq
    provider:
      gke: 
        autopilot: true
  override:
    nodeAgent:
      podOverride: 
        labels:
          provider: datadog
        image: gcr.io/datadoghq/agent:latest-jmx
        containers:
            name: trace-agent: 
            resources:
              requests:
                cpu: 2
            env:
            - name: "DD_DEBUG_LEVEL"
              value: "TRACE"
    clusterAgent: # override for the agent deployment pod
      podOverride: 
        containers: []
[...]
    clusterCheckRunner:
[...]
```

### What does this PR do?

Base to build upon to release the V2 of the operator.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 1]
```diff
@@ -6,28 +6,208 @@
...
+	// Specify additional volumes in the Datadog Agent container.
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// The container image of the Datadog Cluster Checks Runner.
+	Image *ImageConfig `json:"image,omitempty"`
+
+	// If specified, the Agent pod's tolerations.
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:87]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 2]
```diff
@@ -6,28 +6,208 @@
...
+	// are two special keywords which indicate the highest priorities with the former being the highest priority.
+	// Any other name must be defined by creating a PriorityClass object with that name. If not specified,
+	// the pod priority will be default or zero if there is no default.
+	PriorityClassName string `json:"priorityClassName,omitempty"`
+
+	// If specified, the pod's scheduling constraints.
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// AdditionalAnnotations provide annotations that will be added to the Cluster Agent Pods.
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:106]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 3]
```diff
@@ -6,28 +6,208 @@
...
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// AdditionalLabels provide labels that will be added to the Cluster Agent Pods.
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:109]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 4]
```diff
@@ -6,28 +6,208 @@
...
+}
+
+// DatadogAgentGenericContainer is the generic structure describing any container's common configuration.
+// +k8s:openapi-gen=true
+type DatadogAgentGenericContainer struct {
+
+	// Name of the container that is overiden
+	Name string `json:"name,omitempty"`
+
+	// The Datadog Agent supports many environment variables.
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:120]: Todo: update comment + remove link to the doc

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 5]
```diff
@@ -6,28 +6,208 @@
...
+	// See also: https://docs.datadoghq.com/agent/docker/?tab=standard#environment-variables
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Specify additional volume mounts in the APM Agent container.
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:128]:
```suggestion
	// Specify additional volume mounts in the container.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 6]
```diff
@@ -6,28 +6,208 @@
...
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=mountPath
+	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
+
+	// Make sure to keep requests and limits equal to keep the pods in the Guaranteed QoS class.
+	// See also: http://kubernetes.io/docs/user-guide/compute-resources/
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// Command allows the specification of custom entrypoint for Trace Agent container
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:139]:
```suggestion
	// Command allows the specification of custom entrypoint for the container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 7]
```diff
@@ -6,28 +6,208 @@
...
+	// Args allows the specification of extra args to `Command` parameter
+	// +listType=atomic
+	Args []string `json:"args,omitempty"`
+
+	// HealthPort of the agent container for internal liveness probe.
+	// Must be the same as the Liness/Readiness probes.
+	// +optional
+	HealthPort *int32 `json:"healthPort,omitempty"`
+
+	// Configure the Readiness Probe of the Agent container
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:152]:
```suggestion
	// Configure the Readiness Probe of the container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 8]
```diff
@@ -6,28 +6,208 @@
...
+	// +optional
+	ReadinessProbe *corev1.Probe `json:"readinessProbe,omitempty"`
+
+	// Configure the Liveness Probe of the APM container
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:156]:
```suggestion
	// Configure the Liveness Probe of the container
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml [hunk 9]
```diff
@@ -6,22 +6,6 @@ metadata:
   creationTimestamp: null
   name: datadogagents.datadoghq.com
 spec:
-  additionalPrinterColumns:
```
Comment (clamoriniere) [inline config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:9]: this part should have been added back

**Review by clamoriniere — COMMENTED** [review]
Looks good it term of structure definition. The fields descriptions need to be updated

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 10]
```diff
@@ -6,28 +6,208 @@
...
+type ResourceOverride struct {
+	// PodTemplateOverride is used to configure the components at the pod level in an agnotic way
+	PodTemplateOverride *PodTemplateOverride `json:"podOverride,omitempty"`
+
+	// Name overrides the default name for the resource
+	Name string `json:"name,omitempty"`
+}
+
+// PodTemplateOverride is the generic description equivalent to a subset of the PodTemplate for a component
+type PodTemplateOverride struct {
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:74]: Should we name it `DatadogAgentPodTemplateOverride` ?

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 11]
```diff
@@ -6,28 +6,208 @@
...
+	// Registry to use for all Agent images (default gcr.io/datadoghq).
+	// Use public.ecr.aws/datadog for AWS
+	// Use docker.io/datadog for DockerHub
+	// +optional
+	Registry *string `json:"registry,omitempty"`
+}
+
+// ResourceOverride is the generic description of a component (Cluster Agent Deployment, Node Agent Daemonset...)
+// TODO: Add the Resource type and name to allow overriding the kind of the Resource (e.g. ExtendedDaemonset)
+type ResourceOverride struct {
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:65]: Should we rename it as `DatadogAgentResourceOverride`?

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 12]
```diff
@@ -6,28 +6,208 @@
...
 type DatadogAgentSpec struct {
+	// Features running on the Agent and Cluster Agent.
+	// +optional
+	Features *DatadogFeatures `json:"features,omitempty"`
+
+	// Global settings to configure the components
+	// +optional
+	Global *GlobalConfig `json:"global,omitempty"`
+
+	// Override the confiuration of the components
```
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:35]:
```suggestion
	// Override the configuration of the components
```

File: docs/configuration.md [hunk 13]
```diff
@@ -32,393 +32,11 @@ spec:
 
 | Parameter | Description |
 | --------- | ----------- |
-| agent.additionalAnnotations | AdditionalAnnotations provide annotations that will be added to the Agent Pods. |
```
[the remark this answers is not in the document]
Comment (CharlyF) [reply docs/configuration.md]: that's a good point - I will tackle right after we merge this if that's OK

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 14]
```diff
@@ -6,28 +6,209 @@
...
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// ResourceName is the name of a Deployment Component
+type ResourceName string
+
+const (
+	// NodeAgentResourceName is the name of the Datadog Agent
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:17]:
```suggestion
	// NodeAgentResourceName is the name of the Datadog Node Agent
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 15]
```diff
@@ -6,28 +6,209 @@
...
+	NodeAgentResourceName ResourceName = "nodeAgent"
+	// ClusterAgentResourceName is the name of the Cluster Agent
+	ClusterAgentResourceName ResourceName = "clusterAgent"
+	// ClusterCheckRunnerResourceName is the name of the Cluster Check Runner
+	ClusterCheckRunnerResourceName ResourceName = "clusterCheckRunner"
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:22]:
```suggestion
	ClusterChecksRunnerResourceName ResourceName = "clusterChecksRunner"
```

(to be consistent with what we do in Helm https://github.com/DataDog/helm-charts/blob/main/charts/datadog/values.yaml#L1169)

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 16]
```diff
@@ -6,28 +6,209 @@
...
+)
+
 // DatadogAgentSpec defines the desired state of DatadogAgent
 type DatadogAgentSpec struct {
+	// Features running on the Agent and Cluster Agent.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:27]:
```suggestion
	// Features running on the Agent and Cluster Agent
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 17]
```diff
@@ -6,28 +6,209 @@
...
+	// +optional
+	Features *DatadogFeatures `json:"features,omitempty"`
+
+	// Global settings to configure the components
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:31]:
```suggestion
	// Global settings to configure the agents
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 18]
```diff
@@ -6,28 +6,209 @@
...
+	// +optional
+	Global *GlobalConfig `json:"global,omitempty"`
+
+	// Override the configuration of the components
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:35]:
```suggestion
	// Override the default configurations of the agents
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 19]
```diff
@@ -6,28 +6,209 @@
...
+	// Use public.ecr.aws/datadog for AWS
+	// Use docker.io/datadog for DockerHub
+	// +optional
+	Registry *string `json:"registry,omitempty"`
+}
+
+// DatadogAgentResourceOverride is the generic description of a component (Cluster Agent Deployment, Node Agent Daemonset...)
+// TODO: Add the Resource type and name to allow overriding the kind of the Resource (e.g. ExtendedDaemonset)
+type DatadogAgentResourceOverride struct {
+	// DatadogAgentPodTemplateOverride is used to configure the components at the pod level in an agnotic way
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:66]:
```suggestion
	// DatadogAgentPodTemplateOverride is used to configure the components at the pod level in an agnostic way
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 20]
```diff
@@ -6,28 +6,209 @@
...
+	DatadogAgentPodTemplateOverride *DatadogAgentPodTemplateOverride `json:"podOverride,omitempty"`
+
+	// Name overrides the default name for the resource
+	Name string `json:"name,omitempty"`
+}
+
+// DatadogAgentPodTemplateOverride is the generic description equivalent to a subset of the PodTemplate for a component
+type DatadogAgentPodTemplateOverride struct {
+	// Agent config contain the basic configuration of the Datadog Process Agent's container.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:75]:
```suggestion
	// Configure the basic configurations for each agent container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 21]
```diff
@@ -6,28 +6,209 @@
...
+	// Specify additional volumes in the different components (Datadog Agent, Cluster Agent, Cluster Check Runner).
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// The container image of the different components (Datadog Agent, Cluster Agent, Cluster Check Runner).
+	Image *ImageConfig `json:"image,omitempty"`
+
+	// Configure the different components (Datadog Agent, Cluster Agent, Cluster Check Runner) tolerations.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:87]:
```suggestion
	// Configure the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) tolerations.
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml (hunk 9 above)
Comment (clamoriniere) [reply config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:9]: ok,
maybe you could add them (duplicated from the v1alpha1)

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 22]
```diff
@@ -6,28 +6,209 @@
...
+	// are two special keywords which indicate the highest priorities with the former being the highest priority.
+	// Any other name must be defined by creating a PriorityClass object with that name. If not specified,
+	// the pod priority will be default or zero if there is no default.
+	PriorityClassName string `json:"priorityClassName,omitempty"`
+
+	// If specified, the pod's scheduling constraints.
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// Annotations provide annotations that will be added to the different components (Datadog Agent, Cluster Agent, Cluster Check Runner) Pods.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:106]:
```suggestion
	// Annotations provide annotations that will be added to the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) pods.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 23]
```diff
@@ -6,28 +6,209 @@
...
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// AdditionalLabels provide labels that will be added to the different components (Datadog Agent, Cluster Agent, Cluster Check Runner) Pods.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:109]:
```suggestion
	// AdditionalLabels provide labels that will be added to the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) pods.
```

**Review by clamoriniere — APPROVED** [review]
looks good on my side, when others comments will be fixed

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 24]
```diff
@@ -6,28 +6,209 @@
...
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+// DatadogAgentGenericContainer is the generic structure describing any container's common configuration.
+// +k8s:openapi-gen=true
+type DatadogAgentGenericContainer struct {
+
+	// Name of the container that is overiden
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:117]:
```suggestion
	// Name of the container that is overriden
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 25]
```diff
@@ -6,28 +6,209 @@
...
+	//+optional
+	Name string `json:"name,omitempty"`
+
+	// The different components (Datadog Agent, Cluster Agent, Cluster Check Runner) support many environment variables.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:121]: Specify additional environmental variables in the container

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 26]
```diff
@@ -6,28 +6,209 @@
...
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Specify additional volume mounts in the container.
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=mountPath
+	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
+
+	// Make sure to keep requests and limits equal to keep the pods in the Guaranteed QoS class.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:136]: should we write something more generic here? (or, consider having this on a second line and write something generic above it)

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 27]
```diff
@@ -6,28 +6,209 @@
...
+	// See also: http://kubernetes.io/docs/user-guide/compute-resources/
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// Command allows the specification of custom entrypoint for container
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:140]:
```suggestion
	// Command allows the specification of a custom entrypoint for container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 28]
```diff
@@ -6,28 +6,209 @@
...
+	// +listType=atomic
+	Command []string `json:"command,omitempty"`
+
+	// Args allows the specification of extra args to `Command` parameter
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:144]:
```suggestion
	// Args allows the specification of extra args to the `Command` parameter
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 29]
```diff
@@ -6,28 +6,209 @@
...
+	// +listType=atomic
+	Args []string `json:"args,omitempty"`
+
+	// HealthPort of the agent container for internal liveness probe.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:148]:
```suggestion
	// HealthPort of the container for the internal liveness probe.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 30]
```diff
@@ -6,28 +6,209 @@
...
+	// Must be the same as the Liness/Readiness probes.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:149]:
```suggestion
	// Must be the same as the Liveness/Readiness probes.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 31]
```diff
@@ -6,28 +6,209 @@
...
+	// +optional
+	LivenessProbe *corev1.Probe `json:"livenessProbe,omitempty"`
+}
+
+// ImageConfig Datadog agent container image config.
+// +k8s:openapi-gen=true
+type ImageConfig struct {
+	// Define the image to use:
+	// Use "gcr.io/datadoghq/agent:latest" for Datadog Agent 7
+	// Use "datadog/dogstatsd:latest" for Standalone Datadog Agent DogStatsD6
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:167]:
```suggestion
	// Use "datadog/dogstatsd:latest" for standalone Datadog Agent DogStatsD6
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 32]
```diff
@@ -6,28 +6,209 @@
...
+	// Use "gcr.io/datadoghq/cluster-agent:latest" for Datadog Cluster Agent
+	// Use "agent" with the registry and tag configurations for <registry>/agent:<tag>
+	// Use "cluster-agent" with the registry and tag configurations for <registry>/cluster-agent:<tag>
+	Name string `json:"name,omitempty"`
+
+	// Define the image version to use:
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:173]:
```suggestion
	// Define the image tag to use.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 33]
```diff
@@ -6,28 +6,209 @@
...
+	// To be used if the Name field does not correspond to a full image string.
+	// +optional
+	Tag string `json:"tag,omitempty"`
+
+	// Define whether the Agent image should support JMX.
+	// +optional
+	JmxEnabled bool `json:"jmxEnabled,omitempty"`
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:180]: should we use `JMXEnabled`? seems most places we capitalize 'JMX'

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 34]
```diff
@@ -6,28 +6,209 @@
...
+
+	// The Kubernetes pull policy:
+	// Use Always, Never or IfNotPresent.
+	PullPolicy *corev1.PullPolicy `json:"pullPolicy,omitempty"`
+
+	// It is possible to specify docker registry credentials.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:186]:
```suggestion
	// It is possible to specify Docker registry credentials.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 35]
```diff
@@ -6,28 +6,209 @@
...
 	// DefaultOverride contains attributes that were not configured that the runtime defaulted.
 	// +optional
 	DefaultOverride *DatadogAgentSpec `json:"defaultOverride,omitempty"`
 }
 
-//+kubebuilder:object:root=true
-//+kubebuilder:subresource:status
-
-// DatadogAgent is the Schema for the datadogagents API
+// DatadogAgent Deployment with Datadog Operator.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:201]:
```suggestion
// DatadogAgent Deployment with the Datadog Operator.
```

**Comment by celenechang** [comment]
Thanks for doing this. I noticed that for the `nodeAgent`, there is a `podOverride` level above `containers`, but for the `clusterAgent`, `containers` is the next level down. Is this something we could make consistent?

```
   nodeAgent:
      podOverride: 
        containers:
            ...
    clusterAgent:
      containers:
```

**Comment by clamoriniere** [comment]
> Thanks for doing this. I noticed that for the `nodeAgent`, there is a `podOverride` level above `containers`, but for the `clusterAgent`, `containers` is the next level down. Is this something we could make consistent?
> 
> ```
>     nodeAgent:
>       podOverride: 
>         containers:
>             ...
>     clusterAgent:
>       containers:
> ```

Yep the example is not valide. the structure inside `clusterAgent` is  the same than for `nodeAgent`. we can update the PR description

**Review by celenechang — APPROVED** [review]

**Pull request merged**